src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs

```diff
@@ -0,0 +1,454 @@
+        {
+            HttpListenerRequest request = await _helper.GetRequest();
+            using (Stream inputStream = request.InputStream)
+            {
+                Assert.Throws<ArgumentNullException>("buffer", () => inputStream.Read(null, 0, 0));
+                await Assert.ThrowsAsync<ArgumentNullException>("buffer", () => inputStream.ReadAsync(null, 0, 0));
+            }
+        }
+
+        [Theory]
+        [InlineData(-1)]
+        [InlineData(3)]
+        public async Task Read_InvalidOffset_ThrowsArgumentOutOfRangeException(int offset)
+        {
+            HttpListenerRequest request = await _helper.GetRequest();
+            using (Stream inputStream = request.InputStream)
+            {
+                Assert.Throws<ArgumentOutOfRangeException>("offset", () => inputStream.Read(new byte[2], offset, 0));
+                await Assert.ThrowsAsync<ArgumentOutOfRangeException>("offset", () => inputStream.ReadAsync(new byte[2], offset, 0));
+            }
+        }
+
+        [Theory]
+        [InlineData(0, 3)]
+        [InlineData(1, 2)]
+        [InlineData(2, 1)]
+        public async Task Read_InvalidOffsetSize_ThrowsArgumentOutOfRangeException(int offset, int size)
+        {
+            HttpListenerRequest request = await _helper.GetRequest();
+            using (Stream inputStream = request.InputStream)
+            {
+                Assert.Throws<ArgumentOutOfRangeException>("size", () => inputStream.Read(new byte[2], offset, size));
+                await Assert.ThrowsAsync<ArgumentOutOfRangeException>("size", () => inputStream.ReadAsync(new byte[2], offset, size));
+            }
+        }
+
+        [Fact]
+        public async Task EndRead_NullCallback_ThrowsArgumentNullException()
+        {
+            HttpListenerRequest request = await _helper.GetRequest();
+
+            using (Stream inputStream = request.InputStream)
+            {
+                Assert.Throws<ArgumentNullException>("asyncResult", () => inputStream.EndRead(null));
+            }
+        }
+
+        [Fact]
+        public async Task EndRead_InvalidCallback_ThrowsArgumentException()
+        {
+            HttpListenerRequest request1 = await _helper.GetRequest();
+            HttpListenerRequest request2 = await _helper.GetRequest();
+            
+            using (Stream inputStream1 = request1.InputStream)
+            using (Stream inputStream2 = request2.InputStream)
+            {
+                IAsyncResult beginReadResult = inputStream1.BeginRead(new byte[0], 0, 0, new AsyncCallback(Common.StubCallback), null);
+
+                Assert.Throws<ArgumentException>("asyncResult", () => inputStream2.EndRead(new CustomAsyncResult()));
+                Assert.Throws<ArgumentException>("asyncResult", () => inputStream2.EndRead(beginReadResult));
+            }
+        }
+
+        [Fact]
+        public async Task EndRead_CalledTwice_ThrowsInvalidOperationException()
+        {
```

Comment:
This is failing on Unix, due to difference in implemenation. The async result on Unix, doesn't use lazyasyncresult, and we don't check if end was already called. This behavior is not consistent across all Streams. For ex, Network stream doesn't throw for multiple endread calls. cc @stephentoub Should we mimic windows behavior on Unix for this one?